MongoDB Migration release notes for 2023-10 (#166)
ni/install-systemlink-enterprise: release-notes/2023-10/MongoDB_Migration/README.md

# Existing MongoDB Data Migration to a Single Instance

## Do you need to worry about what is in this folder?

SystemLink Enterprise installations prior to 2023-10 deployed a MongoDB instance per service. Version 2023-10 uses a global MongoDB connection string that allows customers to consolidate over 20 different MongoDB instances down to one. This content is for users who need to migrate existing SystemLink Enterprise MongoDB data to a single MongoDB instance. You do not need to perform the following if you are deploying a new installation of SystemLink Enterprise.

## Prerequisites

> Before you begin, you must provision your MongoDB instance with databases and users. Refer to [Configuring Systemlink Enterprise](https://www.ni.com/docs/en-US/bundle/systemlink-enterprise/page/config-systemlink-enterprise.html) for a list of databases and password fields to update if using separate credentials per database. NOTE: Ensure that you backup the current database passwords for the migration.

Reviewer: I'd suggest changing the end of the last sentence to "before the migration."

# Existing MongoDB Data Migration to a Single Instance

## Do you need to worry about what is in this folder?

> The content in this folder can help you if you have recognized a need to migrate existing SystemLink Enterprise MongoDB data to a single MongoDB instance. You should not need this content if deploying a new installation of SystemLink Enterprise.

Reviewer 2: ```suggestion
This content is for users who need to migrate existing SystemLink Enterprise MongoDB data to a single MongoDB instance. You do not need to do this if you are deploying a new installation of SystemLink Enterprise.
```

> ## Background

Reviewer 2: ```suggestion
## Process
```

> Prior to version 2023-10 installations of SystemLink Enterprise deployed a MongoDB instance per service. In version 2023-10 the services were refactored to use a global MongoDB connection string allowing customers to go from running over twenty (20) different MongoDB instances down to one.

Reviewer 2: ```suggestion
SystemLink Enterprise installations prior to 2023-10 deployed a MongoDB instance per service. Version 2023-10 uses a global MongoDB connection string that allows customers to consolidate over 20 different MongoDB instances down to one.
```

Reviewer 2: I would move this context into the "Do you need to worry..." section

> Controlling this functionality is done through global Helm values. Configure `global.mongodb.install` to `false` to enable this feature. You will also need to configure `global.secrets.mongodb.connection_string`. Refer to [Configuring Systemlink Enterprise](https://www.ni.com/docs/en-US/bundle/systemlink-enterprise/page/config-systemlink-enterprise.html) for more information.

Reviewer 2: ```suggestion
Refer to [Configuring Systemlink Enterprise](https://www.ni.com/docs/en-US/bundle/systemlink-enterprise/page/config-systemlink-enterprise.html) for instructions on editing the Helm values that determine this functionality.
```

> ## Prerequisites

Reviewer 2: I would put this info before we send them to the process doc to avoid them running off and missing important info

Author: Do you mean move "Prerequisites" before "Process"?

Reviewer 2: Yeah

Author: I didn't like the "Process" section so I moved some of that content into "Do you need to worry about what is in this folder?" and put the link to ni.com inside a new "Enable a Single Instance" section. Let me know if that flows better.

Reviewer 2: That works for me

> Your MongoDB instance will need to be provisioned with databases and users. Refer to [Configuring Systemlink Enterprise](https://www.ni.com/docs/en-US/bundle/systemlink-enterprise/page/config-systemlink-enterprise.html) for a list of databases. It also lists password fields to update (if using separate credentials per database). NOTE: You will need the current database passwords for the migration. Be sure to backup these values somewhere safe.

Reviewer 2: ```suggestion
Before you begin, you must provision your MongoDB instance with databases and users. Refer to [Configuring Systemlink Enterprise](https://www.ni.com/docs/en-US/bundle/systemlink-enterprise/page/config-systemlink-enterprise.html) for a list of databases and password fields to update if using separate credentials per database. NOTE: Ensure that you backup the current database passwords for the migration.
```

## Migration Plan

### MongoDB Atlas

> If you are using a [MongoDB Atlas](https://www.ni.com/r/mongodbatalas) cluster use MongoDB's Atlas migration tools to move data between the service databases and MongoDB Atlas.

Reviewer 2: ```suggestion
If you are using a [MongoDB Atlas](https://www.ni.com/r/mongodbatalas) cluster use the MongoDB Atlas migration tools to move data between the service databases and MongoDB Atlas.
```

### Self-hosted MongoDB

> From inside the cluster we need to run a `mongodump` piped into a `mongorestore` so that we can migrate the documents to the new MongoDB instance.

Reviewer 2: ```suggestion
Run the following command from inside the cluster to migrate the documents to the new MongoDB instance.
```

> The command we want to run is:

Reviewer 2: ```suggestion
```

```sh
mongodump --config=/etc/mongodump_config.yaml --archive | mongorestore --archive --config=/etc/mongorestore_config.yaml
```

> 

Reviewer 2: ```suggestion
Where
```

> - The `mongo*_config.yaml` files are where the connection string and password is configured.

Reviewer 2: ```suggestion
-  `mongo*_config.yaml` is where the connection string and password are configured.
```

> - `--archive` is used so that we can pipe (`|`) the output directly into the `mongorestore` command

Reviewer 2: ```suggestion
`--archive` allows you to pipe (`|`) the output directly into the `mongorestore` command
```

#### Helm Chart

> Included in this folder is a Helm Chart to install into your existing Kubernetes cluster to perform the above plan. It utilizes a `bitnami/mongodb:5.0.19-debian-11-r3` Docker image which your Kubernetes cluster will need to be able to pull.

Reviewer 2: ```suggestion
This folder includes a Helm Chart to install into your existing Kubernetes cluster to execute the plan described above for self-hosted MongoDB. Your Kubernetes cluster must be able to pull the `bitnami/mongodb:5.0.19-debian-11-r3` Docker image.
```

> You will need to fill out the `values.yaml` file with MongoDB password secrets typically stored in `systemlink-secrets.yaml`. Refer to all `<ATTENTION>` comments.

Reviewer 2: ```suggestion
1. Fill out the `values.yaml` file with MongoDB password secrets stored in `systemlink-secrets.yaml`. Refer to all `<ATTENTION>` comments.
```

> After configuring `values.yaml` install via helm command

Reviewer 2: ```suggestion
```

1. Change active directory on command line to the folder containing this file.
2. Run `helm install ni_systemlink_enterprise_migrate_mongodb .`

You can observe the progress of each command Job through the Pod logs.

> Before continuing ensure all Jobs have completed by running `kubectl get jobs` and observe the `COMPLETIONS` column. `1/1` indicates the Job completed successfully.

Reviewer 2: ```suggestion
Before continuing, run `kubectl get jobs` and observe the `COMPLETIONS` column to ensure all jobs have completed. `1/1` indicates the job completed successfully.
```
